usaspending_api/etl/management/commands/check_es_doc_duplication.py

```diff
@@ -0,0 +1,245 @@
+        self._partition_size = options["partition_size"]
+        es = Elasticsearch(**self._es_client_config)
+        self._index = options["index"]
+        doc_count = es.count(index=options["index"])["count"]
+        _log.info(f"Found {doc_count:,} docs in index {self._index}")
+        self._num_partitions = ceil(doc_count / self._partition_size)
+
+        # Instantiate a numpy array with the number of docs' _id fields we expect to touch
+        all_ids = np.empty(doc_count, dtype=int)
+        _log.info(f"Collecting _ids from all {doc_count:,} in scrolling batch queries of size {self._partition_size:,}")
+
+        # Create a baseline query used in scrolling, which gets all docs, but doesn't return the doc _source.
+        # Use this query in the scan helper, which performs the scrolling
+        # preserve_order=True seems to provide better performance on our indices rather than sort=_doc (see docs)
+        q = {"query": {"match_all": {}}, "_source": False}
+        for idx, hit in enumerate(scan(es, index=self._index, query=q, size=self._partition_size, preserve_order=True)):
+            all_ids[idx] = hit["_id"]
+            doc_num = idx + 1
+            if doc_num % self._partition_size == 0:
+                _log.debug(f"Finished batch {int(doc_num / self._partition_size)} of {self._num_partitions}")
+            percent_complete = floor((doc_num / doc_count) * 100)
+            if percent_complete in self._percents:
+                self._percents.remove(percent_complete)
+                _log.info("=" * ceil(percent_complete * 0.6) + f" {percent_complete}% of scroll complete")
+            if options.get("stop_after") and doc_num >= options["stop_after"]:
+                _log.warning(f"Halting early after reaching --stop-after={options['stop_after']} docs")
+                break
+
+        # Gather stats from _id value array using numpy
+        _log.info("Done scrolling all docs. Gathering statistics on _id values found.")
+        unique, counts = np.unique(all_ids, return_counts=True)
+        duped_ids = unique[counts > 1]  # type: np.ndarray
+        duped_id_counts = counts[counts > 1]  # type: np.ndarray
+        duplicated_doc_ids = dict(zip(duped_ids, duped_id_counts))
+        if 0 in duplicated_doc_ids:
+            duplicated_doc_ids.pop(0)
+
+        if duplicated_doc_ids:
+            duped_id_count = len(duplicated_doc_ids)
+            max_dupe = np.amax(duped_id_counts)
+            p75 = np.percentile(duped_id_counts, 75)
+            p95 = np.percentile(duped_id_counts, 95)
+            _log.warning(
+                f"Found {duped_id_count:,} _ids with more than one doc in the index. "
+                f"Max duplication (p100) = {max_dupe}; p95 = {p95}; p75 = {p75}"
+            )
+            header = "_id         | duplicated\n" + "-" * 12 + "|" + "-" * 12 + "\n"
+            rows = "\n".join([str(k).ljust(12) + "|" + str(v).ljust(12) for k, v in duplicated_doc_ids.items()])
```

Comment:
This printing out of rows is really nice for the `scroll` approach and would be nice to have for the `partition` approach.